service/src/main/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcDocumentTaskExecutor.java

```diff
@@ -43,12 +45,27 @@ public void execute(GetGpcDocumentTaskDefinition taskDefinition) {
         String documentName = taskDefinition.getDocumentId() + JSON_EXTENSION;
         String taskId = taskDefinition.getTaskId();
 
-        var storageDataWrapper = StorageDataWrapperProvider.buildStorageDataWrapper(taskDefinition, response, taskId);
-        storageConnectorService.uploadFile(storageDataWrapper, documentName);
+        var storageDataWrapperWithDocumentRecord = StorageDataWrapperProvider.buildStorageDataWrapper(taskDefinition, response, taskId);
+        storageConnectorService.uploadFile(storageDataWrapperWithDocumentRecord, documentName);
 
-        EhrExtractStatus ehrExtractStatus = ehrExtractStatusService.updateEhrExtractStatusAccessDocument(taskDefinition, documentName,
-            taskId);
+        String messageId = randomIdGeneratorService.createNewId();
+
+        String xmlPayload = translateToXmlPayload(taskDefinition, messageId);
+        var storageDataWrapperWithMhsPayload = StorageDataWrapperProvider.buildStorageDataWrapper(taskDefinition, xmlPayload, taskId);
+        storageConnectorService.uploadFile(storageDataWrapperWithMhsPayload, prepareXmlDocumentName(messageId));
 
+        EhrExtractStatus ehrExtractStatus = ehrExtractStatusService.updateEhrExtractStatusAccessDocument(taskDefinition, documentName,
+            taskId, messageId);
         detectTranslationCompleteService.beginSendingCompleteExtract(ehrExtractStatus);
     }
+
+    private String translateToXmlPayload(GetGpcDocumentTaskDefinition taskDefinition, String messageId) {
```

Comment:
This is only the first part. We now need to take this XML and then serialise it into the JSON structure used for the MHS outbound request. This is because the XML metadata is separate from the binary/base64 attachment. See in the story the file `mhs_outbound_request.json` which describes it.